substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JDK23OrEarlier.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2024, 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core.jdk;
+
+import jdk.graal.compiler.serviceprovider.JavaVersionUtil;
+
+import java.util.function.BooleanSupplier;
+
+public class JDK23OrEarlier implements BooleanSupplier {
```

Comment:
We will never need this. On master, we either support JDK 21 or JDK 24, so `JDK21OrEarlier` and `JDK23OrLater` is sufficient. I plan to rename `JDK23OrLater` to `JDK24OrLater` or maybe even `JDKLatest` to avoid such renamings in the future.

Reply:
>Thanks @roberttoyonaga for your contribution! We have a similar adoption of this change in our JDK update PR https://github.com/oracle/graal/pull/9181. In general, it always makes sense to look for such JDK update PRs when it comes to recent JDK change adoptions and potentially start discussions there.

@zapster Ok I wasn't aware of the updates PR. I'll remember for that for the future. 

 >In our adoption, we also ported the javaTimeSystemUTC method, because it has a higher resolution. (See also my inline comment.) From the only existing usage in jdk.jfr.internal.MetadataRepository#awaitEpochMilliShift, having only millisecond resolution seems fine, which makes me think we should rather use your approach. However, I'm worried that JVM.nanosNow() might be used by other methods in the future, which require sub-millisecond accuracy. Such new usages would be hard to detect. What do you think?

I think porting javaTimeSystemUTC is good. Yes, I think it's fine now, but it would be hard to catch problems if it starts being used for nanosecond accuracy in the future. We should go with your approach.  Thanks!

Reply:
@zapster Actually, on second thought, I think it's important that the chunk timestamps use the same clock as `nanosNow()`. See [the comment here](https://github.com/openjdk/jdk/pull/19799/files#diff-b7f754817501bf15b4cb8004fc066e46fbd97b30b15ef406aa09fd00ca2e5f14R174).  We can keep using `javaTimeSystemUTC`, but I think we ought to also change [how we timestamp chunks](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrChunkFileWriter.java#L166),  probably by changing the clock that backs `JfrTicks.currentTimeNanos()` to `javaTimeSystemUTC`.

Reply:
I opened a new PR for this: https://github.com/oracle/graal/pull/9210

Reply:
sounds reasonable, thank you